src/sympc/protocol/falcon/falcon.py

```diff
@@ -63,3 +66,84 @@ def __eq__(self, other: Any) -> bool:
             return False
 
         return True
+
+    @staticmethod
+    def mul_master(x: MPCTensor, y: MPCTensor, session: Session) -> MPCTensor:
+        """Master method for multiplication.
+
+        Performs Falcon's mul implementation, gets and reshares mul results and distributes shares.
+        This needs to be improved in future, it relies on orchestrator being a trusted third party.
+        Falcon, requires parties to be able to communication between each other.
+
+        Args:
+            x (MPCTensor): Secret
+            y (MPCTensor): Another secret
+            session (Session): Session the tensors belong to
+
+        Returns:
+            shares (ReplicatedSharedTensor): Shares in terms of ReplicatedSharedTensor.
+
+        Raises:
+            ValueError: Raised when number of parties are not three.
+            NotImplementedError: Raised when implementation not present
+
+        """
+        if len(session.parties) != 3:
+            raise ValueError("Falcon requires 3 parties")
+
+        result = None
+
+        if session.protocol.security_type == "semi-honest":
+
+            z_shares = []
+            for party in range(0, len(x.share_ptrs)):
+                z_shares.append(x.share_ptrs[party] * y.share_ptrs[party])
+
+            reshared_shares = []
+
+            def get_shares_and_add_mask(party_rank):
+                przs_mask = (
+                    session.session_ptrs[party_rank]
+                    .przs_generate_random_share(shape=x.shape)
+                    .get_shares()
+                    .get()[0]
+                )
+                share = z_shares[party_rank].get_shares().get()[0] + przs_mask
+                return share
+
+            shares = parallel_execution(get_shares_and_add_mask)([[0], [1], [2]])
+            reshared_shares = ReplicatedSharedTensor.distribute_shares(
+                shares, x.session
+            )
+            result = MPCTensor(shares=reshared_shares, session=x.session)
+
+        elif session.protocol.security_type == "malicious":
+            raise NotImplementedError(
+                "Multiplication has not been implemented for malicious setting"
+            )
+        else:
+            raise NotImplementedError(
+                f"mult operation not implemented for {session.protocol.security_type} setting"
+            )
+
+        return result
+
+    @staticmethod
+    def multiplication_protocol(
+        x: ReplicatedSharedTensor, y: ReplicatedSharedTensor
+    ) -> ReplicatedSharedTensor:
+        """Implementation of Falcon's multiplication with semi-honest security guarantee.
+
+        Args:
+            x (ReplicatedSharedTensor): Secret
+            y (ReplicatedSharedTensor): Another secret
+
+        Returns:
+            shares (ReplicatedSharedTensor): results in terms of ReplicatedSharedTensor.
+        """
+        z_value = (
```

Comment:
MAYBE : we could perform the multiplication,in RSTensor itself,since we are calling the falcon protocol again. Something Like 
<pre><code>op = getattr(operator, "mul")
for x_share, y_share in zip(x.share_ptrs, y.share_ptrs):
z_shares.append(op(x_share, y_share))</code></pre>
we could do it locally , it would return a RSTensor pointer,so we could perform przs masking with round of communication(`get`)

Reply:
Yeah, I wll incorporate this.

Reply:
I tried this, but then RSTensor is not able to access the actual session. The session it accessed has 0 parties so the shares are not being generated. I will try it for a while if I don't find a neat way out. I will create an issue for optimizing this which we can solve next.

Reply:
This was also the reason I had to fall back to checking number of shares rather than number of parties in session.